src/instructlab/model/download.py

```diff
@@ -1,31 +1,258 @@
+        self.download_dest = download_dest
+
+    @abc.abstractmethod
+    def download(self) -> None:
+        """Downloads model from specified repo/release and stores it into download_dest"""
+
+
+class HFDownloader(Downloader):
+    """Class to handle downloading safetensors and GGUF models from Hugging Face"""
+
+    def __init__(
+        self,
+        repository: str,
+        release: str,
+        download_dest: str,
+        filename: str,
+        hf_token: str,
+        ctx,
+    ) -> None:
+        super().__init__(
+            repository=repository, release=release, download_dest=download_dest
+        )
+        self.filename = filename
+        self.hf_token = hf_token
+        self.ctx = ctx
+
+    def download(self):
+        """
+        Download specified model from Hugging Face
+        """
+        click.echo(
+            f"Downloading model from Hugging Face : {self.repository}@{self.release} to {self.download_dest}..."
+        )
+
+        if self.hf_token == "" and "instructlab" not in self.repository:
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to download HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
+
+        try:
+            if self.ctx.obj is not None:
+                hf_logging.set_verbosity(self.ctx.obj.config.general.log_level.upper())
+            files = list_repo_files(repo_id=self.repository, token=self.hf_token)
+            if any(".safetensors" in fname for fname in files):
+                self.download_safetensors()
+            else:
+                self.download_gguf()
+
+        except Exception as exc:
+            click.secho(
+                f"Downloading model failed with the following Hugging Face Hub error: {exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+    def download_gguf(self) -> None:
+        try:
+            hf_hub_download(
+                token=self.hf_token,
+                repo_id=self.repository,
+                revision=self.release,
+                filename=self.filename,
+                local_dir=self.download_dest,
+            )
+
+        except Exception as exc:
+            click.secho(
+                f"Downloading GGUF model failed with the following Hugging Face  Hub error: {exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+    def download_safetensors(self) -> None:
+        try:
+            os.makedirs(
+                name=os.path.join(self.download_dest, self.repository),
+                exist_ok=True,
+            )
+
+            snapshot_download(
+                token=self.hf_token,
+                repo_id=self.repository,
+                revision=self.release,
+                local_dir=os.path.join(self.download_dest, self.repository),
```

Comment:
nit: the local dir is calculated twice (here and in line 114). Could be done once.